### PR TITLE
Windows/Linux keyboard shortcuts at a wordwrap

### LINKED ISSUE #90993: Windows and Linux text navigation shortcuts treat newline and wordwrap differently
On native Windows, shift + home/end will navigate between wordwrapped lines, but not between lines that begin with a newline character.  On Linux, shift+home/end doesn't move between lines at all, whether wordwrap or newline.  On Flutter, we move between lines regardless of wordwrap or newline.

**Native Windows**

https://user-images.githubusercontent.com/389558/135355585-838e867c-b9b3-45ae-9221-9f93c847ce62.mp4

**Flutter on Windows**

https://user-images.githubusercontent.com/389558/135355483-2ffcaec1-c4c8-488e-93d9-941d5b166e65.mov

<details>

<summary>Code</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: const Center(
        child: TextField(
          maxLines: null,
        ),
      ),
    );
  }
}
```

</details>

<details>

<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel windows-shortcuts-missing, 2.6.0-12.0.pre.151, on macOS 11.6 20G165 darwin-x64, locale en-US)
    • Flutter version 2.6.0-12.0.pre.151 at /Users/jmccandless/Projects/flutter
    • Upstream repository git@github.com:justinmc/flutter.git
    • Framework revision b96a131df6 (9 minutes ago), 2021-09-29 15:17:55 -0700
    • Engine revision c9f8cb94ec
    • Dart version 2.15.0 (build 2.15.0-162.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/jmccandless/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 12.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio
    • Android Studio at /Applications/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] VS Code (version 1.60.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.24.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.6 20G165 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 94.0.4606.61

! Doctor found issues in 2 categories.
```

</details>

### PR DESCRIPTION
**Pull request by justinmc**
Windows, Mac, and Linux all have different behavior for how wordwraps are treated by keyboard shortcuts that jump to the beginning/end of a line.  Previously, Flutter always stopped at wordwrapped lines.  This PR aims to get the behavior correct on all platforms.

I believe I covered all of the cases affected by this PR in the tests, so to see exactly what the behavior should be on each platform, check the tests.  These are the high level features:

  * When jumping to the start/end of a line multiple times, some platforms will jump between lines when it's a wordwrap and not a newline.
  * On Mac, home/end scroll to the start/end of the field without moving the cursor.
  * On Mac, shift + home/end selects to the start/end of the document.
  * Bugfix: Some of these kinds of shortcuts weren't calling bringIntoView, so they wouldn't cause the field to scroll.

Fixes https://github.com/flutter/flutter/issues/90993

### CODE REVIEW

**Comment by justinmc** [comment]
I'm going to consider this blocked on https://github.com/flutter/flutter/pull/95948, which will bring back the extend/expand distinction needed to fix my last TODO (shift+home/end expands on Mac).

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 1]
```diff
@@ -9037,7 +9935,7 @@ void main() {
     await sendKeys(
       tester,
       <LogicalKeyboardKey>[
-        LogicalKeyboardKey.home,
+        LogicalKeyboardKey.end,
```
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:9938]: I think this was a typo in this test (probably by me in an earlier PR).

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 2]
```diff
@@ -3445,25 +3472,61 @@ class _WordBoundary extends _TextBoundary {
 // interpreted as caret locations because [TextPainter.getLineAtOffset] is
 // text-affinity-aware.
 class _LineBreak extends _TextBoundary {
-  const _LineBreak(this.textLayout, this.textEditingValue);
+  const _LineBreak(
+    this.textLayout,
+    this.textEditingValue,
+    [this.continuesAtWrap = false]
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:3478]: I think ideally this shouldn't be implemented in `_LineBreak`, as it's only supposed to tell you where the boundary is.

Comment (justinmc) [reply packages/flutter/lib/src/widgets/editable_text.dart]: I've moved it to the action and I think it makes more sense there, thanks for the suggestion.

In order to do this I added continuesAtWrap to DirectionalCaretMovementIntent even though most of its subclasses don't use continuesAtWrap.  Let me know if you think that's not alright.

**Review by LongCatIsLooong — APPROVED** [review]

**Comment by goderbauer** [comment]
(triage) @justinmc Looks like google testing got stuck on this one and I don't see a way to restart it.. Maybe push an empty commit?

**Comment by justinmc** [comment]
I pushed a merge commit, hopefully that fixes it.  Thanks for the reminder.

**Pull request merged**